File: packages/sdkwork-utils-java/src/main/java/com/sdkwork/utils/ObjectUtils.java
package com.sdkwork.utils;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class ObjectUtils {
    private ObjectUtils() {
    }

    public static Map<String, Object> pick(Map<String, Object> source, List<String> keys) {
        Map<String, Object> result = new LinkedHashMap<>();
        for (String key : keys) {
            if (source.containsKey(key)) {
                result.put(key, source.get(key));
            }
        }
        return result;
    }

    public static Map<String, Object> omit(Map<String, Object> source, List<String> keys) {
        Map<String, Object> result = new LinkedHashMap<>();
        for (Map.Entry<String, Object> entry : source.entrySet()) {
            if (!keys.contains(entry.getKey())) {
                result.put(entry.getKey(), entry.getValue());
            }
        }
        return result;
    }

    @SuppressWarnings("unchecked")
    public static Object getPath(Object source, String path) {
        Object current = source;
        for (String segment : splitPath(path)) {
            if (!(current instanceof Map<?, ?> map) || !map.containsKey(segment)) {
                return null;
            }
            current = map.get(segment);
        }
        return current;
    }

    public static boolean hasPath(Object source, String path) {
        Object value = getPath(source, path);
        return value != null;
    }

    @SuppressWarnings("unchecked")
    public static Object shallowMerge(Object base, Object overlay) {
        if (base instanceof Map<?, ?> baseMap && overlay instanceof Map<?, ?> overlayMap) {
            Map<String, Object> result = new LinkedHashMap<>((Map<String, Object>) baseMap);
            for (Map.Entry<?, ?> entry : overlayMap.entrySet()) {
                String key = Objects.toString(entry.getKey(), null);
                if (key != null) {
                    result.put(key, entry.getValue());
                }
            }
            return result;
        }
        return overlay;
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> setPath(Map<String, Object> source, String path, Object value) {
        List<String> segments = splitPath(path);
        if (segments.isEmpty()) {
            return source;
        }

        Map<String, Object> root = new LinkedHashMap<>(source);
        Map<String, Object> current = root;
        for (int index = 0; index < segments.size() - 1; index++) {
            String segment = segments.get(index);
            Object next = current.get(segment);
            Map<String, Object> nextMap;
            if (next instanceof Map<?, ?> existing) {
                nextMap = new LinkedHashMap<>((Map<String, Object>) existing);
            } else {
                nextMap = new LinkedHashMap<>();
            }
            current.put(segment, nextMap);
            current = nextMap;
        }
        current.put(segments.get(segments.size() - 1), value);
        return root;
    }

    @SuppressWarnings("unchecked")
    public static Object deepMerge(Object base, Object overlay) {
        if (base instanceof Map<?, ?> baseMap && overlay instanceof Map<?, ?> overlayMap) {
            Map<String, Object> result = new LinkedHashMap<>((Map<String, Object>) baseMap);
            for (Map.Entry<?, ?> entry : overlayMap.entrySet()) {
                String key = Objects.toString(entry.getKey(), null);
                if (key == null) {
                    continue;
                }
                result.put(
                        key,
                        result.containsKey(key)
                                ? deepMerge(result.get(key), entry.getValue())
                                : entry.getValue());
            }
            return result;
        }
        return overlay;
    }

    public static Map<String, Object> compact(Map<String, Object> source) {
        Map<String, Object> result = new LinkedHashMap<>();
        for (Map.Entry<String, Object> entry : source.entrySet()) {
            if (entry.getValue() != null) {
                result.put(entry.getKey(), entry.getValue());
            }
        }
        return result;
    }

    public static List<String> keys(Map<String, Object> source) {
        return new ArrayList<>(source.keySet());
    }

    public static List<Object> values(Map<String, Object> source) {
        return new ArrayList<>(source.values());
    }

    private static List<String> splitPath(String path) {
        return java.util.Arrays.stream(path.split("\\."))
                .filter(part -> !part.isEmpty())
                .toList();
    }
}

File: packages/sdkwork-utils-kotlin/src/main/kotlin/com/sdkwork/utils/ObjectUtils.kt
package com.sdkwork.utils

object ObjectUtils {
    fun pick(source: Map<String, Any?>, keys: List<String>): Map<String, Any?> {
        val result = LinkedHashMap<String, Any?>()
        for (key in keys) {
            if (source.containsKey(key)) {
                result[key] = source[key]
            }
        }
        return result
    }

    fun omit(source: Map<String, Any?>, keys: List<String>): Map<String, Any?> {
        val result = LinkedHashMap<String, Any?>()
        for ((key, value) in source) {
            if (key !in keys) {
                result[key] = value
            }
        }
        return result
    }

    @Suppress("UNCHECKED_CAST")
    fun getPath(source: Any?, path: String): Any? {
        var current: Any? = source
        for (segment in splitPath(path)) {
            val map = current as? Map<*, *> ?: return null
            if (!map.containsKey(segment)) {
                return null
            }
            current = map[segment]
        }
        return current
    }

    fun hasPath(source: Any?, path: String): Boolean {
        return getPath(source, path) != null
    }

    @Suppress("UNCHECKED_CAST")
    fun shallowMerge(base: Any?, overlay: Any?): Any? {
        if (base is Map<*, *> && overlay is Map<*, *>) {
            val result = LinkedHashMap(base as Map<String, Any?>)
            for ((key, value) in overlay) {
                val stringKey = key?.toString() ?: continue
                result[stringKey] = value
            }
            return result
        }
        return overlay
    }

    @Suppress("UNCHECKED_CAST")
    fun setPath(source: Map<String, Any?>, path: String, value: Any?): Map<String, Any?> {
        val segments = splitPath(path)
        if (segments.isEmpty()) {
            return source
        }

        val root = LinkedHashMap(source)
        var current: MutableMap<String, Any?> = root
        for (index in 0 until segments.size - 1) {
            val segment = segments[index]
            val next = current[segment]
            val nextMap = if (next !is Map<*, *>) {
                LinkedHashMap<String, Any?>()
            } else {
                LinkedHashMap(next as Map<String, Any?>)
            }
            current[segment] = nextMap
            current = nextMap
        }
        current[segments[segments.size - 1]] = value
        return root
    }

    @Suppress("UNCHECKED_CAST")
    fun deepMerge(base: Any?, overlay: Any?): Any? {
        if (base is Map<*, *> && overlay is Map<*, *>) {
            val result = LinkedHashMap(base as Map<String, Any?>)
            for ((key, value) in overlay) {
                val stringKey = key?.toString() ?: continue
                result[stringKey] = if (result.containsKey(stringKey)) {
                    deepMerge(result[stringKey], value)
                } else {
                    value
                }
            }
            return result
        }
        return overlay
    }

    fun compact(source: Map<String, Any?>): Map<String, Any?> {
        return source.filterValues { it != null }
    }

    fun keys(source: Map<String, Any?>): List<String> {
        return source.keys.toList()
    }

    fun values(source: Map<String, Any?>): List<Any?> {
        return source.values.toList()
    }

    private fun splitPath(path: String): List<String> {
        return path.split(".")
            .filter { it.isNotEmpty() }
    }
}
